pkg/controllers/nodeclaim/tagging/controller.go

```diff
@@ -0,0 +1,136 @@
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/samber/lo"
+
+	"github.com/aws/karpenter/pkg/apis/v1beta1"
+	"github.com/aws/karpenter/pkg/providers/instance"
+	"github.com/aws/karpenter/pkg/utils"
+
+	corev1beta1 "github.com/aws/karpenter-core/pkg/apis/v1beta1"
+	"github.com/aws/karpenter-core/pkg/cloudprovider"
+	corecontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+)
+
+const (
+	TagNodeClaim = corev1beta1.Group + "/nodeclaim"
+	TagName      = "Name"
+)
+
+type Controller struct {
+	kubeClient       client.Client
+	instanceProvider *instance.Provider
+}
+
+func NewController(kubeClient client.Client, instanceProvider *instance.Provider) corecontroller.Controller {
+	return corecontroller.Typed[*corev1beta1.NodeClaim](kubeClient, &Controller{
+		kubeClient:       kubeClient,
+		instanceProvider: instanceProvider,
+	})
+}
+
+func (c *Controller) Name() string {
+	return "nodeclaim.tagging"
+}
+
+func (c *Controller) Reconcile(ctx context.Context, nodeClaim *corev1beta1.NodeClaim) (reconcile.Result, error) {
+	storedNodeClaim := nodeClaim.DeepCopy()
+
+	if !isTaggable(nodeClaim) {
+		return reconcile.Result{}, nil
+	}
+
+	id, err := utils.ParseInstanceID(nodeClaim.Status.ProviderID)
+	if err != nil {
+		// We don't throw an error here since we don't want to retry until the ProviderID has been updated.
+		logging.FromContext(ctx).Errorf("failed to parse instance ID, %w", err)
+		return reconcile.Result{}, nil
+	}
+
+	err = c.tagInstance(ctx, nodeClaim, id)
+	if err != nil {
+		if !cloudprovider.IsNodeClaimNotFoundError(err) {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	}
+
+	nodeClaim.Annotations = lo.Assign(nodeClaim.Annotations, map[string]string{v1beta1.AnnotationInstanceTagged: "true"})
+	if !equality.Semantic.DeepEqual(nodeClaim, storedNodeClaim) {
+		if err := c.kubeClient.Patch(ctx, nodeClaim, client.MergeFrom(storedNodeClaim)); err != nil {
+			return reconcile.Result{}, client.IgnoreNotFound(err)
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Builder(_ context.Context, m manager.Manager) corecontroller.Builder {
+	return corecontroller.Adapt(
+		controllerruntime.
+			NewControllerManagedBy(m).
+			For(&corev1beta1.NodeClaim{}).
+			WithEventFilter(predicate.NewPredicateFuncs(func(o client.Object) bool {
+				nc := o.(*corev1beta1.NodeClaim)
+				return isTaggable(nc)
+			})),
+	)
+}
+
+func (c *Controller) tagInstance(ctx context.Context, nc *corev1beta1.NodeClaim, id string) error {
+	tags := map[string]string{
+		TagName:      nc.Status.NodeName,
+		TagNodeClaim: nc.Name,
+	}
+
+	// Remove tags which have been already populated
+	instance, err := c.instanceProvider.Get(ctx, id)
+	if err != nil {
+		return err
+	}
+	tags = lo.OmitByKeys(tags, lo.Keys(instance.Tags))
+
+	if len(tags) != 0 {
+		err = c.instanceProvider.CreateTags(ctx, id, tags)
+		time.Sleep(time.Second)
```

Comment:
Is there a better solution here and have we seen this happening in testing? It strikes me as an unnecessary precaution and slowness for users that may not run into the rate limiting here.

Reply:
I haven't tested this controller at scale yet but it would very likely be an issue in a large scale-up. The reason we're being cautious here is because `CreateTags` shares a bucket with other mutating API calls and we'd rather this be slow than get rate limited on other, more important operations (e.g. `CreateFleet`). It would be good to get some data to see if this would actually be an issue in practice though.

Reply:
Can we try to rate-limit this as part of the underlying queue mechanism rather than injecting a sleep into our code?

Reply:
To my understanding the rate limiter is only used for retries, not for overall requests. If there is a way to get it working like that I absolutely agree that would be better.

Reply:
Yeah, after discussing with @jonathan-innis, it seems like this the sleep is the best way to do this 👍